### PR TITLE
Semantics of structure constructors, part 2 of 2

### PR DESCRIPTION
**Pull request by klausler**
Using the changes recently merged into `lib/evaluate`, these changes address most of the remaining TODO items for semantic analysis of structure constructors in expressions.

Several weeks' worth of other small changes have also accumulated here during development.

### CODE REVIEW

**Comment by klausler** [comment]
Be advised, `test/semantics/resolve51.f90` fails with these changes.  Tim's work in progress to analyze the interfaces of contained subprograms before performing the semantic checking of host executable parts will fix that failure.

File: lib/semantics/assignment.cc [hunk 1]
```diff
@@ -25,9 +26,218 @@
...
+  if (auto proc{Characterize(f.proc(), intrinsics)}) {
+    std::optional<parser::MessageFixedText> error;
+    if (const auto &funcResult{proc->functionResult}) {
+      const auto *frProc{funcResult->IsProcedurePointer()};
+      if (IsProcedurePointer(lhs)) {
+        // Shouldn't be here in this function unless lhs
+        // is an object pointer.
+        error = "Procedure pointer '%s' assigned with result of "
+                "reference to function '%s' that does not return a "
+                "procedure pointer"_err_en_US;
```
Comment (tskeith) [inline lib/semantics/assignment.cc]: "Assigned with" sounds odd to me. How about "associated with"?

Comment (klausler) [reply lib/semantics/assignment.cc]: "was assigned the result" may be even better

**Comment by sscalpone** [comment]
The jenkins build/test complains about three tests. The following tests FAILED:
	 79 - resolve51.f90 (Failed)
	112 - allocate11.f90 (Failed)
	117 - null01.f90 (Failed)
@klausler Are the failures besides resolve51 expected?
@tskeith Are you changes to address the issue coming soon?

**Comment by klausler** [comment]
allocate11 and null01 are new tests.  null01 passed before i merged.  I'll look into them.

**Comment by jeanPerier** [comment]
allocate11 failures comes from `symbol.get<DerivedTypeDetails>()` on a symbol that does not have `DerivedTypeDetails` in `const Symbol *HasCoarrayUltimateComponent(const DerivedTypeSpec &derivedTypeSpec)` at `lib/semantics/tools.cc:318`.
The symbol is `derivedTypeSpec.typeSymbol()`. Shouldn't these kind of symbols always have a `DerivedTypeDetails`?

**Pull request merged**